### PR TITLE
add lightning colab tutorial notebooks

### PR DESCRIPTION
**Pull request by nateraw**
## What does this PR do?

Reprise of #2929

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by nateraw** [comment]
@Borda how do I skip the notebooks when building docs? Mind helpin me out there? seems they're causing docs build to fail.

File: notebooks/01-mnist-hello-world.ipynb [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "view-in-github",
+        "colab_type": "text"
+      },
+      "source": [
+        "<a href=\"https://colab.research.google.com/github/nateraw/pytorch-lightning/blob/colab-tuts/notebooks/01_mnist_hello_world.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
Comment (nateraw) [inline notebooks/01-mnist-hello-world.ipynb:26]:
```suggestion
        "<a href=\"https://colab.research.google.com/github/PytorchLightning/pytorch-lightning/blob/master/notebooks/01-mnist-hello-world.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```

File: notebooks/02-datamodules.ipynb [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "view-in-github",
+        "colab_type": "text"
+      },
+      "source": [
+        "<a href=\"https://colab.research.google.com/github/nateraw/pytorch-lightning/blob/colab-tuts/notebooks/02_datamodules.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
Comment (nateraw) [inline notebooks/02-datamodules.ipynb:26]:
```suggestion
        "<a href=\"https://colab.research.google.com/github/PytorchLightning/pytorch-lightning/blob/master/notebooks/02-datamodules.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```

File: notebooks/03-basic-gan.ipynb [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "view-in-github",
+        "colab_type": "text"
+      },
+      "source": [
+        "<a href=\"https://colab.research.google.com/github/nateraw/pytorch-lightning/blob/colab-tuts/notebooks/03-basic-gan.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
Comment (nateraw) [inline notebooks/03-basic-gan.ipynb:25]:
```suggestion
        "<a href=\"https://colab.research.google.com/github/PytorchLightning/pytorch-lightning/blob/master/notebooks/03-basic-gan.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```

File: notebooks/04-transformers-text-classification.ipynb [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "view-in-github",
+        "colab_type": "text"
+      },
+      "source": [
+        "<a href=\"https://colab.research.google.com/github/nateraw/pytorch-lightning/blob/colab-tuts/notebooks/04-transformers-text-classification.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
Comment (nateraw) [inline notebooks/04-transformers-text-classification.ipynb:26]:
```suggestion
        "<a href=\"https://colab.research.google.com/github/PytorchLightning/pytorch-lightning/blob/master/notebooks/04-transformers-text-classification.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```

**Review by nateraw — COMMENTED** [review]
gotta commit these so the links work correctly

**Comment by nateraw** [comment]
@edenafek I plan on getting these merged and opening 2nd PR to fix links in documentation to tutorials, as I want to be sure the links work before pasting them everywhere. Otherwise, if they're broken after we merge, we'll have to do the same boring work twice. I'd like to avoid that 😄 .

**Comment by Borda** [comment]
> @Borda how do I skip the notebooks when building docs? Mind helpin me out there? seems they're causing docs build to fail.

https://github.com/PyTorchLightning/pytorch-lightning/blob/6309a0dbb85fbe249f5c29130c82369aa9ae3a2b/docs/source/conf.py#L306-L312

**Comment by Borda** [comment]
just checking the compatibility with COlab and jupytext
- https://colab.research.google.com/github/PyTorchLightning/pytorch-lightning/blob/add-notebooks/notebooks/01-mnist-hello-world.ipynb
- https://colab.research.google.com/github/PyTorchLightning/pytorch-lightning/blob/add-notebooks/notebooks/01-mnist-hello-world.py

**Review by Borda — APPROVED** [review]

**Comment by nateraw** [comment]
> just checking the compatibility with Colab and `jupytext`
> 
> * https://colab.research.google.com/github/PyTorchLightning/pytorch-lightning/blob/add-notebooks/notebooks/01-mnist-hello-world.ipynb
> * https://colab.research.google.com/github/PyTorchLightning/pytorch-lightning/blob/add-notebooks/notebooks/01-mnist-hello-world.py

Hey @Borda, @teddykoker advised me to remove the .py files for now since we didn't get jupytext workin with colab. So I did that.

**Review by teddykoker — APPROVED** [review]

File: notebooks/01-mnist-hello-world.ipynb [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+      },
+      "source": [
+        "# Introduction to Pytorch Lightning ⚡\n",
+        "\n",
+        "In this notebook, we'll go over the basics of lightning by preparing models to train on the [MNIST Handwritten Digits dataset](https://en.wikipedia.org/wiki/MNIST_database).\n",
+        "\n",
+        "---\n",
+        "  - Give us a ⭐ [on Github](https://www.github.com/PytorchLightning/pytorch-lightning/)\n",
+        "  - Check out [the documentation](https://pytorch-lightning.readthedocs.io/en/latest/)\n",
+        "  - Join us [on Slack](https://join.slack.com/t/pytorch-lightning/shared_invite/zt-f6bl2l0l-JYMK3tbAgAmGRrlNr00f1A)"
```
Comment (edenlightning) [inline notebooks/01-mnist-hello-world.ipynb:43]: Ask a question in our forum (https://forums.pytorchlightning.ai/)

Comment (nateraw) [reply notebooks/01-mnist-hello-world.ipynb:43]: Yes!

File: notebooks/01-mnist-hello-world.ipynb [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+        "colab": {}
+      },
+      "source": [
+        "class MNISTModel(pl.LightningModule):\n",
+        "\n",
+        "    def __init__(self):\n",
+        "        super(MNISTModel, self).__init__()\n",
+        "        self.l1 = torch.nn.Linear(28 * 28, 10)\n",
+        "\n",
+        "    def forward(self, x):\n",
```
Comment (edenlightning) [inline notebooks/01-mnist-hello-world.ipynb:120]: I think the current approach is to move away from adding forward loops. @williamFalcon

File: notebooks/01-mnist-hello-world.ipynb (hunk 5 above)
Comment (nateraw) [inline notebooks/01-mnist-hello-world.ipynb:43]:
```suggestion
        "  - Join us [on Slack](https://join.slack.com/t/pytorch-lightning/shared_invite/zt-f6bl2l0l-JYMK3tbAgAmGRrlNr00f1A)"
        "  - Ask a question in [our forum](https://forums.pytorchlightning.ai/)"
```

File: notebooks/01-mnist-hello-world.ipynb [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "EHpyMPKFkVbZ",
+        "colab_type": "text"
+      },
+      "source": [
+        "## Simplest example\n",
+        "\n",
+        "Here's the simplest most minimal example with just a training loop (no validation, no testing).\n",
```
Comment (edenlightning) [inline notebooks/01-mnist-hello-world.ipynb:101]: mention running on GPU

File: notebooks/03-basic-gan.ipynb [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+        "\n",
+        "Main takeaways:\n",
+        "1. Generator and discriminator are arbitraty PyTorch modules.\n",
+        "2. training_step does both the generator and discriminator training.\n",
+        "\n",
+        "---\n",
+        "\n",
+        "  - Give us a ⭐ [on Github](https://www.github.com/PytorchLightning/pytorch-lightning/)\n",
+        "  - Check out [the documentation](https://pytorch-lightning.readthedocs.io/en/latest/)\n",
+        "  - Join us [on Slack](https://join.slack.com/t/pytorch-lightning/shared_invite/zt-f6bl2l0l-JYMK3tbAgAmGRrlNr00f1A)"
```
Comment (nateraw) [inline notebooks/03-basic-gan.ipynb:47]:
```suggestion
        "  - Join us [on Slack](https://join.slack.com/t/pytorch-lightning/shared_invite/zt-f6bl2l0l-JYMK3tbAgAmGRrlNr00f1A)"
        "  - Ask a question in [our forum](https://forums.pytorchlightning.ai/)"
```

File: notebooks/04-transformers-text-classification.ipynb [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+        "# Finetune 🤗 Transformers Models with PyTorch Lightning ⚡\n",
+        "\n",
+        "This notebook will use HuggingFace's `datasets` library to get data, which will be wrapped in a `LightningDataModule`. Then, we write a class to perform text classification on any dataset from the[ GLUE Benchmark](https://gluebenchmark.com/). (We just show CoLA and MRPC due to constraint on compute/disk)\n",
+        "\n",
+        "[HuggingFace's NLP Viewer](https://huggingface.co/nlp/viewer/?dataset=glue&config=cola) can help you get a feel for the two datasets we will use and what tasks they are solving for.\n",
+        "\n",
+        "---\n",
+        "  - Give us a ⭐ [on Github](https://www.github.com/PytorchLightning/pytorch-lightning/)\n",
+        "  - Check out [the documentation](https://pytorch-lightning.readthedocs.io/en/latest/)\n",
+        "  - Join us [on Slack](https://join.slack.com/t/pytorch-lightning/shared_invite/zt-f6bl2l0l-JYMK3tbAgAmGRrlNr00f1A)\n",
```
Comment (nateraw) [inline notebooks/04-transformers-text-classification.ipynb:45]:
```suggestion
        "  - Join us [on Slack](https://join.slack.com/t/pytorch-lightning/shared_invite/zt-f6bl2l0l-JYMK3tbAgAmGRrlNr00f1A)"
        "  - Ask a question in [our forum](https://forums.pytorchlightning.ai/)\n",
```

File: notebooks/02-datamodules.ipynb [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+        "\n",
+        "This notebook will walk you through how to start using Datamodules.\n",
+        "\n",
+        "The most up to date documentation on datamodules can be found [here](https://pytorch-lightning.readthedocs.io/en/latest/datamodules.html).\n",
+        "\n",
+        "---\n",
+        "\n",
+        "  - Give us a ⭐ [on Github](https://www.github.com/PytorchLightning/pytorch-lightning/)\n",
+        "  - Check out [the documentation](https://pytorch-lightning.readthedocs.io/en/latest/)\n",
+        "  - Join us [on Slack](https://join.slack.com/t/pytorch-lightning/shared_invite/zt-f6bl2l0l-JYMK3tbAgAmGRrlNr00f1A)"
```
Comment (nateraw) [inline notebooks/02-datamodules.ipynb:48]:
```suggestion
        "  - Join us [on Slack](https://join.slack.com/t/pytorch-lightning/shared_invite/zt-f6bl2l0l-JYMK3tbAgAmGRrlNr00f1A)"
        "  - Ask a question in [our forum](https://forums.pytorchlightning.ai/)"
```

File: notebooks/01-mnist-hello-world.ipynb [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+        "        super(MNISTModel, self).__init__()\n",
+        "        self.l1 = torch.nn.Linear(28 * 28, 10)\n",
+        "\n",
+        "    def forward(self, x):\n",
+        "        return torch.relu(self.l1(x.view(x.size(0), -1)))\n",
+        "\n",
+        "    def training_step(self, batch, batch_nb):\n",
+        "        x, y = batch\n",
+        "        loss = F.cross_entropy(self(x), y)\n",
+        "        return pl.TrainResult(loss)\n",
```
Comment (edenlightning) [inline notebooks/01-mnist-hello-world.ipynb:126]: maybe we can simply return the loss here, and explain the trainResult object in the second example?

**Review by nateraw — COMMENTED** [review]
thank you @edenafek

**Review by edenlightning — APPROVED** [review]

File: notebooks/01-mnist-hello-world.ipynb (hunk 6 above)
Comment (nateraw) [reply notebooks/01-mnist-hello-world.ipynb:120]: really?

File: notebooks/01-mnist-hello-world.ipynb (hunk 11 above)
Comment (nateraw) [reply notebooks/01-mnist-hello-world.ipynb:127]: good eye..might be good idea. I'll do a little more cleanup after initial merge. don't want to mess with json files too much right now so I avoid breaking things.

**Comment by nateraw** [comment]
adding comment not to merge as I see if links still work w/ changes I just commited

**Comment by Borda** [comment]
@nateraw "don't merge" is not blocking command, WIP is =)
